From bcf96e82c0d340556a4ed863f76788510f002555 Mon Sep 17 00:00:00 2001
From: Jeremy Drake <github@jdrake.com>
Date: Thu, 10 Sep 2020 10:31:34 -0700
Subject: [PATCH 3/3] Don't crash if asked to create an empty implib.

With --enable-reloc-section (now default), ld segfaults if asked to
create an import lib for an executable with 0 exports.  With
--disable-reloc-section (the old default), it would just silently not
create one.  This patch restores the latter behavior for the
--enable-reloc-section case.

PR 26588
---
 ld/emultempl/pe.em  | 3 ++-
 ld/emultempl/pep.em | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/ld/emultempl/pe.em b/ld/emultempl/pe.em
index b3990cf677..32cd175ebb 100644
--- a/ld/emultempl/pe.em
+++ b/ld/emultempl/pe.em
@@ -2012,7 +2012,8 @@ gld_${EMULATION_NAME}_finish (void)
     )
     {
       pe_dll_fill_sections (link_info.output_bfd, &link_info);
-      if (command_line.out_implib_filename)
+      if (command_line.out_implib_filename
+          && pe_def_file->num_exports != 0)
 	pe_dll_generate_implib (pe_def_file, command_line.out_implib_filename,
 				&link_info);
     }
diff --git a/ld/emultempl/pep.em b/ld/emultempl/pep.em
index de6d7aafbb..0b43ec515d 100644
--- a/ld/emultempl/pep.em
+++ b/ld/emultempl/pep.em
@@ -1865,7 +1865,8 @@ gld_${EMULATION_NAME}_finish (void)
 	  && pep_def_file->num_exports != 0))
     {
       pep_dll_fill_sections (link_info.output_bfd, &link_info);
-      if (command_line.out_implib_filename)
+      if (command_line.out_implib_filename
+          && pep_def_file->num_exports != 0)
 	pep_dll_generate_implib (pep_def_file,
 				 command_line.out_implib_filename, &link_info);
     }
-- 
2.28.0.windows.1

